### PR TITLE
Add boolean suppression DebugSessionOptions

### LINKED ISSUE #12015: [vscode] Support DebugSessionOptions boolean options for suppression
see https://code.visualstudio.com/api/references/vscode-api#DebugSessionOptions

4 new options since 1.73:

- suppressDebugStatusbar?: boolean
_When true, the window statusbar color will not be changed for this session._

- suppressDebugToolbar?: boolean
_When true, the debug toolbar will not be shown for this session._

- suppressDebugView?: boolean
_When true, the debug viewlet will not be automatically revealed for this session._

- suppressSaveBeforeStart?: boolean
_When true, a save will not be triggered for open editors when starting a debug session, regardless of the value of the debug.saveBeforeStart setting._

### PR DESCRIPTION
**Pull request by sgraband**
#### What it does

Add options to suppress the debugging statusbar, toolbar and view. 
As well as the saving before starting a debugging session.

Fixes #12015 

Contributed on behalf of STMicroelectronics

_Note:_
Parent/child sessions are currently handled as followed:
- For the saving supression only the root session is taken into account as this could have unintended side effects (saving files when not wanted), if a child session specifies something else than the parent session.
- For the other three supressions (debug view, toolbar and statusbar) the value of the current session is evaluated. If there is no value the nearest parent session will be evaluated.

I mostly went with two different approaches here due to the significance of the actions (saving vs graphical-only), but your input is welcome and i will gladly unify the approaches, when the "better" solution is identified.

#### How to test

- Install this extension: [helloworld-sample-0.0.1.zip](https://github.com/eclipse-theia/theia/files/10811670/helloworld-sample-0.0.1.zip)
It registers a few commands that start the "Launch Browser Backend" debug config with different options set.
- Open theia and the theia repository as the workspace.
- Add the following to your setting.json file:

>   "debug.saveBeforeStart": "allEditorsInActiveGroup",
>   "debug.openDebug": "openOnSessionStart",
>   "files.autoSave": "off",

- Trigger one of the `Start Debugging` commands. It says in parenthesis which options are enabled.
- See if the behavior is correct:

> suppressDebugStatusbar?: boolean
> When true, the window statusbar color will not be changed for this session.
> 
> suppressDebugToolbar?: boolean
> When true, the debug toolbar will not be shown for this session.
> 
> suppressDebugView?: boolean
> When true, the debug viewlet will not be automatically revealed for this session.
> 
> suppressSaveBeforeStart?: boolean
> When true, a save will not be triggered for open editors when starting a debug session, regardless of the value of the debug.saveBeforeStart setting.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/debug/src/browser/debug-session-manager.ts [hunk 1]
```diff
@@ -210,7 +210,8 @@ export class DebugSessionManager {
     protected async startConfiguration(options: DebugConfigurationSessionOptions): Promise<DebugSession | undefined> {
         return this.progressService.withProgress('Start...', 'debug', async () => {
             try {
-                if (!await this.saveAll()) {
+                // If a parent session is available saving should be handled by the parent
+                if (!options.configuration.parentSessionId && !options.configuration.suppressSaveBeforeStart && !await this.saveAll()) {
```
Comment (tsmaeder) [inline packages/debug/src/browser/debug-session-manager.ts:214]: This seems to be a drive-by fix. Was our behaviour broken before? If so, what was the bug?

File: packages/debug/src/common/debug-configuration.ts [hunk 2]
```diff
@@ -72,6 +72,21 @@ export interface DebugConfiguration {
...
     /** Task to run after debug session ends */
     postDebugTask?: string | TaskIdentifier;
+
+    /**
+     * When true, a save will not be triggered for open editors when starting a debug session,
+     * regardless of the value of the `debug.saveBeforeStart` setting.
+     */
+    suppressSaveBeforeStart?: boolean;
+
+    /** When true, the debug toolbar will not be shown for this session. */
```
Comment (tsmaeder) [inline packages/debug/src/common/debug-configuration.ts:82]: As mentioned above, I don't think this flag makes sense in Theia. If Theia does not use it, It shouldn't appear in Theia code.

**Review by tsmaeder — COMMENTED** [review]
A couple of questions/suggestions

File: packages/debug/src/browser/debug-frontend-application-contribution.ts [hunk 3]
```diff
@@ -1494,10 +1496,36 @@ export class DebugFrontendApplicationContribution extends AbstractViewContributi
...
+            } else {
+                break;
+            }
+        }
+        return result;
+    }
+
+    protected shouldViewBeSuppressed(session: DebugSession): boolean | undefined {
+            // Only suppress the view if both the debug view and the toolbar are suppressed, as the toolbar is always docked.
+            return this.getOption(session, 'suppressDebugView') && this.getOption(session, 'suppressDebugToolbar');
```
[the remark this answers is not in the document]
Comment (sgraband) [reply packages/debug/src/browser/debug-frontend-application-contribution.ts:1529]: Mhm.. I see your point. My reasoning behind this was that if the extensions sets `suppressDebugView: true` but does not set `suppressDebugToolbar` they might expect to see the toolbar. 

But you are probably right, that we should just ignore the `suppressDebugToolbar` all together. I will change this.

File: packages/debug/src/browser/debug-session-manager.ts (hunk 1 above)
Comment (sgraband) [reply packages/debug/src/browser/debug-session-manager.ts:214]: No this is basically just what i described here:

> Note:
Parent/child sessions are currently handled as followed:

> For the saving supression only the root session is taken into account as this could have unintended side effects (saving files when not wanted), if a child session specifies something else than the parent session.

I wanted to make sure, that if the config that is executed specifies `suppressSaveBeforeStart: true` the save is never executed, e.g. if child configs do not specify the flag. WDYT?

**Comment by tsmaeder** [comment]
There does not seem to be an option in the settings UI for `debug.saveBeforeStart`. Is this maybe missing in the preference schema?

**Review by tsmaeder — CHANGES_REQUESTED** [review]
I'm fine with the changes, but we need to decide whether the support for the `saveOnStart` preference should be in this PR or not. If not, could you please open an issue so we can track the problem?

**Comment by sgraband** [comment]
I created #12319 for the follow up.

**Review by tsmaeder — APPROVED** [review]
Looks good to me now.
